### ISSUE TITLE
npm install not working in npm 6.9.0 when authentication changes

### ISSUE THREAD

**Issue by sathishchandran97** [verdaccio/verdaccio#1381]
**Describe the bug**
npm install is not working for npm v6.9.0 when the access is change from $all to particular users in config.yaml

**To Reproduce**
1. In server: Go to 'config.yaml'

2. In server: Under packages section grant the 'access' to one/many users (space separated) instead of $all.
    Something like
```
packages:
'@*/*':
# scoped packages
access: xxx yyy
publish: xxx
unpublish: xxx
proxy: npmjs
```

3. In server: Optionally enable logs with 
```
logs:
# - { type: stdout, format: pretty, level: info, path: verdaccio.log}
- {type: file, path: verdaccio.log, level: info}
```

4. In client: Login with username xxx or yyy. Login is successful

5. In client: run npm install

6. You will see
```
npm WARN @scope/package-name@0.1.7 No repository field.

npm ERR! code E401
npm ERR! Unable to authenticate, your authentication token seems to
be invalid.
npm ERR! To correct this please trying logging in again with:
npm ERR!     npm login

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\User\AppData\Roaming\npm-cache\_logs\2019-07-09T14_38_41_734Z-debug.log
```

**Expected behavior**
It should work and the package should be downloaded into node_modules as there is no problem if npm version is v3.10.10 (verified)

**Update**
I have set this up(verdaccio) in my local machine running npm v6.9.0 and I tried running npm install in different machine in the same LAN. It works with both npm 3.10.10 and 6.9.0. May be this happens only in server which is running centos7(could be the reason) or verdaccio in external network (possibilities are there) . Hope this helps!

[2019-07-09T14_38_41_734Z-debug.log](https://github.com/verdaccio/verdaccio/files/3373645/2019-07-09T14_38_41_734Z-debug.log)

**Comment by juanpicado** [comment]
Could you format it please? I think some context is missing.

**Comment by sathishchandran97** [comment]
> Could you format it please? I think some context is missing.

yeah did it

**Comment by sathishchandran97** [comment]
@juanpicado  I hope these four log files would help you

Logs from verdaccio running in server (centos7). Hence the clients here are not in same LAN. 

In the file centos7-server-user-sathish.txt **_make a note that the user is null_**
[centos7-server-user-sathish.txt](https://github.com/verdaccio/verdaccio/files/3377086/centos7-server-user-sathish.txt)
[centos7-server-user-jagan.txt](https://github.com/verdaccio/verdaccio/files/3377087/centos7-server-user-jagan.txt)




Logs from verdaccio running in local (windows). Hence the clients here are from the same LAN.

Everything works perfect here.
[windows-local-user-sathish.txt](https://github.com/verdaccio/verdaccio/files/3377091/windows-local-user-sathish.txt)
[windows-local-user-jagan.txt](https://github.com/verdaccio/verdaccio/files/3377092/windows-local-user-jagan.txt)

**Comment by juanpicado** [comment]
@sathishchandran97 have you tried to look into it? or tried with other toolking as `yarn`, `pnpm`?

**Comment by sathishchandran97** [comment]
> @sathishchandran97 have you tried to look into it? or tried with other toolking as `yarn`, `pnpm`?

yeah @juanpicado you are right. It is working well with yarn add package-name, but fails with npm install package-name

**Comment by juanpicado** [comment]
@sathishchandran97 :-) ask them why does not work, perhaps dig into the npm cli.

I'd rather ask them here https://npm.community/  , the issue might be in their side or our side, but if yarn works, the most of the time has been in their side.

**Comment by sathishchandran97** [comment]
> I'd rather ask them here https://npm.community/ , the issue might be in their side or our side, but if yarn works, the most of the time has been in their side.

ok. thanks @juanpicado for your time. would follow up and let you know :)

**Comment by sathishchandran97** [comment]
@juanpicado Just one small help. Please go through these 2 logs. 

log 1: 
![difference1](https://user-images.githubusercontent.com/50132149/61053065-673a1a00-a40a-11e9-8e26-2415eab3930e.PNG)

log 2: 
![difference2](https://user-images.githubusercontent.com/50132149/61053715-b5035200-a40b-11e9-9f16-cb9634d243f9.PNG)

If you can see the above two screenshots,
1. There are additional fields 'pacote-req-type' and 'pacote-pkg-id' are added in npm v6.9.0. Does verdaccio do this or it is by npm ??
2. There are additional fields 'npm-in-ci' and 'npm-scope' in npm v6.9.0. Again the same question as above.
3. In the final object the user suddenly becomes 'null' and status becomes '401' in npm v6.9.0. Again the same question as above.

Please respond when you find time. Thanks in advance :)

**Comment by juanpicado** [comment]
Perhaps you can tell me what's the source of each log, some context, I'm a bit confused.

> There are additional fields 'pacote-req-type' and 'pacote-pkg-id' are added in npm v6.9.0. Does verdaccio do this or it is by npm ??

npmjs

> There are additional fields 'npm-in-ci' and 'npm-scope' in npm v6.9.0. Again the same question as above

npmjs

> In the final object the user suddenly becomes 'null' and status becomes '401' in npm v6.9.0. Again the same question as above.

I'm not sure where I have to look here, which user?

**Comment by sathishchandran97** [comment]
Both the pictures are taken from same log file by enabling 
```
logs:
- {type: file, path: verdaccio.log, level: info
```
The one to the left in both the pictures is created when a machine which uses npm v6.9.0 runs npm install and to the right is created when a machine  uses npm v3.10.10 runs npm install. Both these machines (clients) use the same user name 'atom' to download the package

**Comment by juanpicado** [comment]
Ok, I see, well, we do not modify any of those private / custom metadata fields, actually all pass through. I don't know what that `user` field might be. What I do not see in the read field, is the `authorization` header, which is the field that contains the TOKEN and that might be the reason of the 401, no token, no access.

**Comment by DanielRuf** [comment]
Does a `npm publish` through npmjs work? Maybe it is an issue with `.npmrc`.

Can you try again after `npm login`?

**Comment by sathishchandran97** [comment]
@DanielRuf Except npm install, all other commands are working well. The only issue is when we give npm install, it shows E401 .authorization tokens seems to be invalid error

**Comment by sathishchandran97** [comment]
This issue got fixed by adding `always-auth=true` in .npmrc file.
To do so run `npm config set always-auth=true` from cmd.

Thanks @juanpicado and @DanielRuf for your time and support.

**Comment by juanpicado** [comment]
Oh ... :-) Thanks for the feedback. Interesting.

**Comment by kopax** [comment]
For the records:

> npm config set always-auth=true

I was struggling for hours to understand why I had the `401` when doing `npm @myscope/package --registry https://npm.myserver.com` within a node project.

As soon as I have typed `npm config set always-auth=true` I got in mp `~/.npmrc` a line with `always-auth=true`, then the command was succeeding.

I then removed from the `.npmrc` the new line and redo again the installation and then it was still working! (I am positive nothing change except I have done that command and removed it's result, then it worked normally).

It only work without it for that package, every other secured packages will fail, this is why I must keep the `always-auth=true` in the `.npmrc`.

From NPM documentation:

> ### `always-auth§`
> - Default: false
> - Type: Boolean
> Force npm to always require authentication when accessing the registry, even for GET requests.

I don't see any hint here that could explain such behavior, if anyone have the explanation, thanks in advance for sharing it when you have time.
